File: OJ/leetcode/python/51_n-queens.py
def solveNQueens(n):
    def isValid(i, j, col, diagonal_down, diagonal_up):
        return col[j] and diagonal_up[i + j] \
            and diagonal_down[len(col) - i - 1 + j]

    def dfs(i, diagonal_down, diagonal_up, col, board, ans, n):
        if i == n:
            ans.append(["".join(line) for line in board])
            return
        
        for j in range(n):
            if isValid(i, j, col, diagonal_down, diagonal_up):
                col[j] = 0
                diagonal_up[i + j] = 0
                diagonal_down[n - i - 1 + j] = 0
                board[i][j] = "Q"
                dfs(i + 1, diagonal_down, diagonal_up, col, board, ans, n)
                board[i][j] = "."
                diagonal_down[n - i - 1 + j] = 1
                diagonal_up[i + j] = 1
                col[j] = 1

    col = [1 for _ in range(n)]
    # diagonal from left-top to right-down 
    diagonal_down = [1 for _ in range(2 * n - 1)]
    # diagonal from left-down to right-up 
    diagonal_up = [1 for _ in range(2 * n - 1)]
    board = [["." for _ in range(n)] for _ in range(n)]
    ans = []
    dfs(0, diagonal_down, diagonal_up, col, board, ans, n)
    return ans

n = 5
ans = solveNQueens(n)
for line in ans:
    print("=" * n)
    for row in line:
        print(row)

File: competitive_programming/leetcode/cpp/51_n-queens.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

bool isValid(int x, int y, 
             const vector<bool>& col,
             const vector<bool>& down,
             const vector<bool>& up) {
    return col[y] && up[x + y] && down[col.size() - 1 - x + y];
}

void dfs(int x, 
         vector<bool>& col, vector<bool>& down, vector<bool>& up,
         vector<string>& board, vector<vector<string>>& ans) {
    if (x == board.size()) {
        ans.push_back(board);
        return;
    }

    for (int y = 0; y < board.size(); ++y) {
        if (isValid(x, y, col, down, up)) {
            board[x][y] = 'Q';
            col[y] = false;
            down[board.size() - 1 - x + y] = false;
            up[x + y] = false;
            dfs(x + 1, col, down, up, board, ans);
            up[x + y] = true;
            down[board.size() - 1 - x + y] = true;
            col[y] = true;
            board[x][y] = '.';
        }
    }
}

vector<vector<string>> solveNQueens(int n) {
    vector<string> board(n, string(n, '.'));
    vector<bool> col(n, true);
    // from top-left to down-right
    vector<bool> down(2 * n - 1, true);
    // from down-left to top-right
    vector<bool> up(2 * n - 1, true);

    vector<vector<string>> ans;
    dfs(0, col, down, up, board, ans);
    return ans;
}

void print(const vector<vector<string>>& ans) {
    for (vector<string> line : ans) {
        for (string s : line) {
            cout << s << "\n";
        }
    }
    cout << ans.size();
}

int main() {
    vector<vector<string>> ans;
    ans = solveNQueens(8);
    print(ans);
    return 0;
}

